File: finished/java/museum-exhibit-studio/src/main/java/workshop/CuratorStreamer.java
package workshop;

import com.github.copilot.CopilotSession;
import com.github.copilot.generated.AssistantMessageDeltaEvent;
import com.github.copilot.generated.AssistantMessageEvent;
import com.github.copilot.generated.ToolExecutionCompleteEvent;
import com.github.copilot.generated.ToolExecutionStartEvent;
import com.github.copilot.rpc.MessageOptions;

import java.time.Duration;
import java.util.concurrent.atomic.AtomicBoolean;

public final class CuratorStreamer {
    public static final Duration GENERATION_TIMEOUT = Duration.ofSeconds(120);
    public static final Duration RESEARCH_TIMEOUT = Duration.ofSeconds(90);

    private CuratorStreamer() {
    }

    public static String streamExhibit(CopilotSession session, String prompt) throws Exception {
        return streamExhibit(session, prompt, GENERATION_TIMEOUT);
    }

    public static String streamExhibit(CopilotSession session, String prompt, Duration timeout) throws Exception {
        if (timeout.isNegative() || timeout.toMillis() == 0) {
            throw new IllegalArgumentException("The response timeout must be positive.");
        }
        StringBuffer assistantText = new StringBuffer();
        AtomicBoolean receivedDelta = new AtomicBoolean(false);
        try (var subscription = session.on(event -> {
            if (event instanceof AssistantMessageDeltaEvent delta && delta.getData() != null) {
                String content = delta.getData().deltaContent();
                if (content != null && !content.isEmpty()) {
                    receivedDelta.set(true);
                    assistantText.append(content);
                    System.out.print(content);
                }
            } else if (event instanceof AssistantMessageEvent message && message.getData() != null) {
                String content = message.getData().content();
                if (!receivedDelta.getAndSet(false) && content != null && !content.isEmpty()) {
                    assistantText.append(content);
                    System.out.print(content);
                }
            } else if (event instanceof ToolExecutionStartEvent tool && tool.getData() != null) {
                System.out.println("\n[tool:start] " + tool.getData().toolName());
            } else if (event instanceof ToolExecutionCompleteEvent tool && tool.getData() != null) {
                System.out.println("[tool:done] success=" + tool.getData().success());
            }
        })) {
            // The SDK wait handles session.error, idle, send failures and timeout, and releases its listener.
            var response = session.sendAndWait(new MessageOptions().setPrompt(prompt), timeout.toMillis());
            try {
                response.get();
            } finally {
                response.cancel(true);
            }
            System.out.println();
            return assistantText.toString();
        }
    }
}

File: start-museum/dotnet/Helpers/CuratorStreamer.cs
using System.Text;
using GitHub.Copilot;

namespace MuseumExhibitStudio.Helpers;

public static class CuratorStreamer
{
    public static readonly TimeSpan GenerationTimeout = TimeSpan.FromSeconds(120);
    public static readonly TimeSpan ResearchTimeout = TimeSpan.FromSeconds(90);

    public static async Task<string> StreamExhibitAsync(
        CopilotSession session,
        string prompt,
        TimeSpan? timeout = null,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(session);
        ArgumentNullException.ThrowIfNull(prompt);

        var response = new StringBuilder();
        var completed = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
        var receivedDelta = false;
        var actualTimeout = timeout ?? GenerationTimeout;
        if (actualTimeout <= TimeSpan.Zero)
        {
            throw new ArgumentOutOfRangeException(nameof(timeout), "The response timeout must be positive.");
        }
        using var deadline = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        deadline.CancelAfter(actualTimeout);

        using var subscription = session.On<SessionEvent>(sessionEvent =>
        {
            if (completed.Task.IsCompleted)
            {
                return;
            }
            switch (sessionEvent)
            {
                case AssistantMessageDeltaEvent delta when !string.IsNullOrEmpty(delta.Data.DeltaContent):
                    receivedDelta = true;
                    response.Append(delta.Data.DeltaContent);
                    Console.Write(delta.Data.DeltaContent);
                    break;
                case AssistantMessageEvent message:
                    if (!receivedDelta && !string.IsNullOrEmpty(message.Data.Content))
                    {
                        response.Append(message.Data.Content);
                        Console.Write(message.Data.Content);
                    }
                    receivedDelta = false;
                    break;
                case ToolExecutionStartEvent tool:
                    Console.WriteLine($"\n[tool:start] {tool.Data.ToolName}");
                    break;
                case ToolExecutionCompleteEvent tool:
                    Console.WriteLine($"[tool:done] success={tool.Data.Success}");
                    break;
                case SessionIdleEvent:
                    Console.WriteLine();
                    completed.TrySetResult();
                    break;
                case SessionErrorEvent error:
                    completed.TrySetException(new InvalidOperationException(error.Data.Message));
                    break;
            }
        });

        try
        {
            var send = session.SendAsync(new MessageOptions { Prompt = prompt }, deadline.Token);
            var first = await Task.WhenAny(send, completed.Task).WaitAsync(deadline.Token);
            await first;
            await send.WaitAsync(deadline.Token);
            await completed.Task.WaitAsync(deadline.Token);
        }
        catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
        {
            throw new TimeoutException("The curator response reached the timeout.");
        }
        finally
        {
            await deadline.CancelAsync();
        }
        return response.ToString();
    }
}
